Optionally set SPDX `licenseConcluded` to ORT's effective license

sschuberth opened the pull request:
Previously, only ORT's `concludedLicense` from a package curation was taken into account. However, if solely detected license findings were cleared via license finding curations from package configurations, that did not have any impact at all until now.

To fix this, use the effective license with a custom license view of the declared and detected licenses as a fallback if no concluded license is set.

Finally, the `licenseConcluded` should only be set if (human) clearance work was involved, so only set it if it differs from the `licenseDeclared`. If `licenseDeclared` already was correct from the start, then this needs to be "acknowledged" by manually setting the concluded license to the same expression.

sschuberth committed:
feat(spdx): Optionally set `licenseConcluded` to the effective license

Previously, only ORT's `concludedLicense` from a package curation was
taken into account. However, if solely detected license findings were
cleared via license finding curations from package configurations, that
did not have any impact at all until now.

To address this, use the effective license with a custom license view of
the declared and detected licenses as a fallback if no concluded license
is set.

Finally, the `licenseConcluded` should only be set if (human) clearance
work was involved, so only set it if it differs from the
`licenseDeclared`. If `licenseDeclared` already was correct from the
start, then this needs to be "acknowledged" by manually setting the
concluded license to the same expression.

Signed-off-by: Sebastian Schuberth <sebastian@doubleopen.org>

sschuberth committed:
feat(spdx): Optionally set `licenseConcluded` to the effective license

Previously, only ORT's `concludedLicense` from a package curation was
taken into account. However, if solely detected license findings were
cleared via license finding curations from package configurations, that
did not have any impact at all until now.

To address this, use the effective license with a custom license view of
the declared and detected licenses as a fallback if no concluded license
is set.

Finally, the `licenseConcluded` should only be set if (human) clearance
work was involved, so only set it if it differs from the
`licenseDeclared`. If `licenseDeclared` already was correct from the
start, then this needs to be "acknowledged" by manually setting the
concluded license to the same expression.

Signed-off-by: Sebastian Schuberth <sebastian@doubleopen.org>

tsteenbe commented on plugins/reporters/spdx/src/main/kotlin/Extensions.kt:
```diff
@@ -159,6 +160,14 @@ internal fun Package.toSpdxPackage(
         .applyChoices(ortResult.getPackageLicenseChoices(id))
         .applyChoices(ortResult.getRepositoryLicenseChoices())
 
+    val licenseDeclared = resolvedLicenseInfo.mainLicense()?.simplify()
```
> We have no mechanism to signal that "a human has finished looking at all license findings and made curations as necessary"

That question is a process question, not a SBOM one imo. I believe we should not change the concluded license mechanism at package level. Concluded license curations within package configurations should be transcoded as concluded license on file level e.g. https://spdx.github.io/spdx-spec/v2.3/file-information/.

If you want to communicate whether a process has finished in a SBOM then I would suggest you initiated a conversation with [SPDX Operations Profile](https://spdx.github.io/spdx-spec/v3.1-RC1/model/Operations/Operations/) folks.

By automatically setting concluded license via ORT you taking away a user choice to intentionally provide no concluded license information to their SBOM consumers. A lot of organizations I know/work with have legal counsel that deliberately want the concluded license field in SPDX SBOMs to be a NOASSERTION as setting may result in contractual obligations. 

If you need this behavior for your organization then build a ORT plugin for this but please do not make this part of core ORT functionality.

sschuberth replied:
> we could put this behind an opt-in feature toggle

Done so now.

tsteenbe replied:
Thinking about this PR some more it makes even more no sense we implementing a change to concluded license behavior for SPDX reporter only which would result ORT giving inconsistent concluded license reporting per package depending on which report a users looks whether SPDX, CycloneDX, WebApp or notices. When reports have inconsistent data between each other the trust in ORT to do a decent job will be diminished. Futuremore, having such a option would make traceability of data origin and training of people in ORT license concepts more difficult - let alone the fact that it may be going again SPDX definition of a concluded license being "Contain the license the SPDX document creator has concluded as governing the package". 

Don't think this PR can be quickly solved by creating a "autoConcludeToEffectiveLicense" option but say we discuss this in one of our next calls. Would like to know what is the real user use case we trying to fix here?

sschuberth committed:
refactor(spdx): Pull out a variable for later reuse

This keeps the diff smaller for an upcoming change.

Signed-off-by: Sebastian Schuberth <sebastian@doubleopen.org>

sschuberth committed:
refactor(spdx): Pull out a variable for later reuse

This keeps the diff smaller for an upcoming change.

Signed-off-by: Sebastian Schuberth <sebastian@doubleopen.org>

sschuberth committed:
chore(spdx): Reorder config options to make output formats go last

Somewhat try to use "chronological" order to options, i.e. the order in
which they are used by the implementation in a report workflow.

Signed-off-by: Sebastian Schuberth <sebastian@doubleopen.org>

sschuberth committed:
chore(spdx): Reorder config options to make output formats go last

Somewhat try to use "chronological" order to options, i.e. the order in
which they are used by the implementation in a report workflow.

Signed-off-by: Sebastian Schuberth <sebastian@doubleopen.org>

arohen-junlegal commented:
Adding a legal compliance practitioner perspective, since the discussion has been mostly tooling-centric so far.

The technical concerns raised here are well taken, and we follow the reasoning. From a legal standpoint, however, we do need a populated licenseConcluded field and it is, after all, what SPDX provides this field for. Downstream SBOM ingestion systems in the automotive supply chain treat license identifiers per subcomponent as mandatory, both at package and at file level. Where clearance is done exclusively via license finding curations which, as @tsteenbe rightly notes, is the cleaner ORT approach. There is currently no way to surface that result in SPDX without misusing package-level concluded-license curations as a workaround.

Ideally, this PR would address both levels:
• Package level: the aggregated effective license, as proposed here.
• File level: the per-file curated result, transcoded into SPDX licenseConcluded on file information along the lines @tsteenbe already suggested above. For our use case, the file-level mapping is at least as important as the package-level one, since downstream consumers rely on per-subcomponent identifiers.
A general point that may be worth stating explicitly, since it underlies several of the concerns raised:

the role of the tool here is to provide a standards-conformant way of expressing curation results in SPDX. Whether a given curated result may, in a concrete deployment, be emitted as licenseConcluded is a legal and organizational decision of the entity running the scan; informed by its contractual situation, internal clearance process, and counsel. ORT enabling the mapping does not commit anyone to using it; an opt-in toggle (off by default) makes that division of responsibility explicit. Seen from this angle, most of the concerns below are arguments about defaults and documentation, not about whether the capability should exist in core.
On the specific concerns:
• We have separately discussed declared vs. concluded with an SPDX maintainer, and the reading reflected in this PR (declared = upstream repo, concluded = post-curation result) was confirmed.
• The contractual-obligation concern is legitimate but, per the above, handled by the opt-in default.
• The reporter-consistency point is fair; promoting this to a general reporter option rather than SPDX-only would resolve it.
A plugin-only solution would push users toward forks for what is a defensible reading of an existing SPDX field. Opt-in in core covering both package and file level seems like the right balance.

DwRene commented:
I would also like to comment on this topic, but rather from a technical perspective. I am speaking here from a support standpoint and regarding the issues we very frequently face concerning the discussion of licenseDeclared and licenseConcluded.

Since we also accept SPDX 2.3 data for our legal compliance tool, we adhere to the SPDX 2.3 standard, which states that licenseConcluded contains the final license determination made by the creator of the SBOM.

Package level: "Contains the license the SPDX document creator has concluded as governing the package or alternative values, if the governing license cannot be determined."

File level: "This field contains the license the SPDX document creator has concluded as governing the file or alternative values if the governing license cannot be determined."

This is crucial for a legal compliance tool, as it involves legal assessments regarding the use of open-source software. From a technical perspective, as a tool developer, I need to know where the correct information is located in the SBOM I receive so that it can ultimately be stored in the right place for license clearance within a tool. If everyone stores the data according to their own interpretation and perhaps adds explanations in free-text fields, you can imagine the chaos this leads to.

I can understand very well that from an ORT perspective—meaning at the tool level—there might not be a final standardized decision because the tool cannot make this decision on behalf of the open-source software user. Furthermore, any user can install and run ORT locally, which makes a global solution difficult.

Therefore, I see two problems here:

1. User Awareness: I feel that many ORT users are not aware of the fact that they must perform a final curation or license determination if they want to submit their SBOM for a legal compliance check. We repeatedly hear the argument: "I created and exported the data with ORT in SPDX 2.3 format, so it is correct."

2. Automation Blockers: Automation is severely bottlenecked if you cannot write the data to the licenseConcluded field at both the package and file level. 

For us, this implies potentially having to use an additional post-processing tool.

From this point of view, it would be much better to have an opt-in option in ORT to export the data with the licenseConcluded fields populated based on the effective license.
